### PR TITLE
Patch to support xNVMe

### PR DESCRIPTION
**Pull request by Baekalfen**
Hi, I used autopxd over at [xNVMe](https://github.com/OpenMPDK/xNVMe), and needed to patch a few things to make it work, and I thought it could be of use to others.

The commits should be self-contained and have each the functionality they describe.

An overview:
 * `--regex` on command-line to allow arbitrary conversions. Some datatypes or syntax just doesn't get parsed correctly by autopxd. As a quick solution, I can hotfix something easily without making full and proper support in autopxd -- for example an opaque pointer in a function I don't need. (`--regex 's/FILE \*/void */g'`).
 * Flattening members of unnamed unions and structs. Cython just cannot express these, so I made a "hack" where these members become accessible ~~at least in some manner~~. See example below.
 * Make enums into `cpdef` to make them accessible from Python as well as Cython for free.
 * Ignore StaticAssert. Simply skip them. They don't make much sense in Cython and currently it makes autopxd output faulty code.
 * ~~Possible to ignore known/defined declarations. Maybe there's a better way to do this, but I needed this to support multiple `.h` files. This way, I can provide the declarations from other headers, so they are not repeated.~~ There is a better way. The `whitelist` kwarg.
 * Replace `wget` in favor of `urllib`. Not all distros has `wget` (nor `cURL`) installed, so I opted for a native Python implementation. This works across all distros I have tried ([xNVMe test run](https://github.com/OpenMPDK/xNVMe/actions/runs/2313928575)).

I would love to get feedback from you, especially on the part I will highlight.

```C
struct test_struct {
    int a;
    struct {
        int b;
        int c;
    }
}
```

```Cython
cdef struct test_struct:
    int a
    int b
    int c
```

### CODE REVIEW

File: autopxd/__init__.py [hunk 1]
```diff
@@ -64,22 +65,37 @@ def preprocess(code, extra_cpp_args=None, debug=False):
...
     decls = []
     for decl in ast.ext:
-        if not hasattr(decl, "name") or decl.name not in IGNORE_DECLARATIONS:
+        if hasattr(decl, "type") and hasattr(decl.type, "name") and decl.type.name in additional_ignore_declarations:
+            continue
+
+        if hasattr(decl, "decl") and hasattr(decl.decl, "name") and decl.decl.name in additional_ignore_declarations:
+            continue
+
+        if not hasattr(decl, "name") or decl.name not in (IGNORE_DECLARATIONS | additional_ignore_declarations):
```
Comment (Baekalfen) [inline autopxd/__init__.py]: I would love to hear, if there's a better way to do this.

**Comment by elijahr** [comment]
@Baekalfen this is great! Can you get tests passing?

**Comment by Baekalfen** [comment]
> @Baekalfen this is great! Can you get tests passing?

Yes, I'll get it fixed. I realize the thing with anon nested structs (like below) isn't totally correct. I'll investigate further if it needs to be narrowed down or removed.

```C
    int nested__d
    int nested__e
```

**Comment by Baekalfen** [comment]
They should be fixed now to get the best of both worlds.  The confusing came from trying to print the structs.

 This will work for any struct as long as it only has **named** nested structs inside it.
 ```Python
 cdef test_struct A = ...
 A.nested.x = 2
 print(A)
 ```

 But the same setup with a struct that contains a nested anon struct will result in a compile error along the lines of:
 ```
 test_cython.c:4240:18: error: conflicting types for '__pyx_convert__to_py_struct___xnvme_opts_css'
 static PyObject* __pyx_convert__to_py_struct___xnvme_opts_css(struct _xnvme_opts_css s) {
 ```

 But dropping the `print` of the entire struct, will make it work. I believe this is either a bug in Cython or an accepted limitation from their side.

 ```Python
 cdef test_struct A = ...
 A.nested.x = 2
 print(A.a)
 print(A.b)
 print(A.c)
 print(A.nested.x)
 ```

**Comment by Baekalfen** [comment]
@elijahr The tests and the formatting should all be done now, but there's something with the linting step in GHA.

**Comment by Baekalfen** [comment]
I removed the additional ignore thing. Using the `whitelist` kwarg was much easier once I got it to work. I think the docs are a little off. You have to add `'<stdin>'` to the whitelist to allow declarations from the `.h` file you are providing. Otherwise it only takes declarations from the whitelisted includes.

**Comment by elijahr** [comment]
This is great! Thanks so much for the contribution!

**Comment by elijahr** [comment]
@Baekalfen this has been released in https://pypi.org/project/autopxd2/2.1.1/

**Comment by Baekalfen** [comment]
> @Baekalfen this has been released in https://pypi.org/project/autopxd2/2.1.1/

Fantastic! Thank you!

**Pull request merged**